core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,246 @@
+use super::{
+    hasher::{self, Digest},
+    AdviceSetError, Word,
+};
+use math::fields::f64::BaseElement;
+use std::collections::HashMap;
+
+// MERKLE PATH SET
+// ================================================================================================
+
+/// A map of Merkle Paths
+///
+/// This struct is intended to be used as one of the variants of the MerkleSet enum.
+#[derive(Clone, Debug)]
+pub struct MerklePathSet {
+    root: Word,
+    total_depth: u32,
+    paths: HashMap<u64, Vec<Word>>,
+}
+
+impl MerklePathSet {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an empty MerklePathSet.
+    pub fn new(depth: u32) -> Result<Self, AdviceSetError> {
+        let root = [BaseElement::new(0); 4];
+        let paths = HashMap::<u64, Vec<Word>>::new();
+
+        Ok(Self {
+            root,
+            total_depth: depth,
+            paths,
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds the Merkle Path to the paths map.
+    pub fn add_path(
+        &mut self,
+        depth: u32,
+        index: u64,
+        path: Vec<Word>,
+    ) -> Result<(), AdviceSetError> {
+        let index = 2u64.pow(depth) + index / 2;
+        self.paths.insert(index, path);
+        Ok(())
+    }
+
+    /// Returns the root of parent Merkle tree.
+    pub fn root(&self) -> Word {
+        self.root
+    }
+
+    /// Returns the depth of this Merkle tree.
+    ///
+    /// Merkle tree of depth 1 has two leaves, depth 2 has four leaves etc.
+    pub fn depth(&self) -> u32 {
+        self.total_depth
+    }
+
+    /// Returns a node at the specified depth and index.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the specified depth.
+    ///
+    /// Q: Now I can only return the hash of the node whose index is stored in the map. But I also
+    /// have hashes of other vertices that are present in the path. Should I be able to return
+    /// them too?
```

Comment:
For now, let's allow only requesting nodes at the leaf level. That is, the passed-in depth must be the same as the depth of of the Merkle path set.